Provide a Docker image

gassmoeller opened the pull request:
Following the recent discussion about docker containers on the mailing list, I thought it would be nice to have a more practical discussion, including an example for how such a image could look like, and how the workflow of running a model would look like. When reading into the topic again I noticed some improvements on Docker's side since I last read about it (~1-2 years ago). Installation on Ubuntu is now better explained, and some comfort features I was missing before have been added (like cleanup of finished containers and dangling images: `docker container prune` and `docker image prune`). Unfortunately they introduced the cool new Docker for Windows app that only runs on Win 10 Professional, and hide the legacy solution for Win 10 Home pretty well (it is called docker toolbox, if anyone is looking for it). Nevertheless, also docker-toolbox for Win10 Home is pretty easy to use.

With an installed docker engine installing and running aspect is super simple:
`docker pull gassmoeller/aspect` will download the image
`docker run -it gassmoeller/aspect bash` will start a shell in the container, with precompiled aspect and ready to do. Mounting a volume into the container is explained in the manual entry I wrote.

Things we should discuss:
1. Do we want to offer this option? This should be answered after discussing the remaining questions.
2. Is my description of a typical use case in the manual accurate? I remember several occasions, where I had 1-4 hrs to sit together with somebody and discuss how to run their problem in ASPECT. In each of these cases I would have loved to have the docker image, because it would have allowed to set up the user to further develop their model. So far I either: 1. created an input file on my machine, and after the session send them the file, more or less saying: Figure out how you install ASPECT on your machine, then you can use this file. Or 2. Download the virtual image if there is a fast enough internet connection, and set them up in the virtual machine (with the disadvantage described on the mailing list, that one does not usually like to work in a different desktop environment than one is used to).
3. Is the current setup a reasonable one? Should we rely on the deal.II docker files, or offer our own solution? Building on deal.II requires less work on our side, but as far as I understood the dealii files are not officially supported either, so they might change, or be removed at some point. 
4. For whatever reason the current Dockerfile does not automatically build on docker hub. As far as I can tell we do not violate their build limitations (1CPU, 2GB RAM, 2hrs build time), but for the moment I need to manually rebuild the image. I could do this on a regular basis (e.g. 1/day) to keep the image up-to-date with the master branch.
5. In the current form I only build ASPECT in release mode. If we decide to offer the image we should probably include compiled debug and release versions with the same advice as always: test debug mode, before running release mode.

Although I was skeptical 2 years ago, I have to say I think the docker option is an alternative for many users. Not necessarily for tutorial usage, but more for users with installation problems (I spent several days installing aspect on OpenSUSE, and I know of many people having issues with older MacOS versions, all of those would benefit from this option).

tjhei commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,87 @@ \subsection{Particles}
+../aspect your_input_file.prm
+\end{verbatim}
+
+\item After the model has finished (or during the model run if you want to check intermediate results) the model output will appear in the mounted directory. You can visualize the data outside of the container in your favorite visualization program. Warning: The files are accessible both from within the container and from your host operating system. Try to avoid concurrent write processes.
```
Is concurrent writing really an issue and something we need to discuss? I can not think of a case where it matters.

tjhei commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,87 @@ \subsection{Particles}
+\item After the model has finished (or during the model run if you want to check intermediate results) the model output will appear in the mounted directory. You can visualize the data outside of the container in your favorite visualization program. Warning: The files are accessible both from within the container and from your host operating system. Try to avoid concurrent write processes.
+
+\begin{verbatim}
+docker container prune
```
what is this block about?

tjhei commented on docker/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+          -DDEAL_II_DIR=$HOME/deal.II-install \
+          .. && \
+    make -j4 >log 2>&1 && \
+    mv aspect $HOME/aspect/ && \
```
why mv and not cp or a symlink? I would do the latter.

hlokavarapu replied:
Hi Rene, I just built the image on my local laptop and noticed that the build info is being redirected to a log file

`make -j4 >log 2>&1`

which is then discarded.

This is only a suggestion but I think that it is informative to know the status of the build, whether it is 10% done or 90% done and so on. However, I may be alone on this.

gassmoeller replied:
I redirected the output, because gcc 6.2 currently creates hundreds of deprecation warnings while compiling aspect (if I remember correctly there was an issue within gcc, that should be fixed in newer versions), so the output is very confusing. I am fine with removing this when this is fixed at some point. Anyway, I think most users will not need to build the image themselves, because it is available at docker hub.

tjhei commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,87 @@ \subsection{Particles}
+\item Create your \aspect{} input file in a folder of your choice (possibly also containing any input model data that is required by your model).
+\item Run the docker image and mount the prepared directory as a volume into the docker container. The code below assumes you have navigated into the prepared directory. When you have started the container run the aspect model inside the container, and make sure the model output is saved in your prepared directory.
+\begin{verbatim}
+docker run -it -v "$(pwd):/home/dealii/aspect/model" \
```
you fail to mention the main problem with mounting directories: file permission and ownership is a big issue.

tjhei replied:
I think it would also be required to talk about the following here:
1. how you stop/delete/start a container again
2. everything outside that directory will be lost

gassmoeller replied:
I am a bit confused. These points seem not longer as critical as they used to be (at least I did not encounter them), but we can certainly discuss them in the manual. 
On my machine when mounting the directory it appears with uid and permissions for the docker user specified in the dockerfile ('dealii') inside of the container. From the outside all files in that directory (newly created or changed) have my (host) user's permissions and uid, so somehow docker now translates these uids and permissions (or is this by accident because dealii and my host user happen to have the same uid?). When I accidentally mount a directory that belongs to root, it also appears as root inside of the container, and the dealii user is not allowed to change those. It is also not possible to run su or sudo from within the container, therefore you can not accidentally break things or change permissions. I do not know about mounting other users directories, but can give it a try tomorrow. I am not sure if this depends on the type of docker storage driver (I use Aufs under ubuntu), but they seem to handle file permissions pretty well nowadays. Of course in a user created container the user could make himself root and do crazy things, but in our container with a prescribed non-root user without sudo-privileges it seems safe.
Addition: Just tested windows and had some issues getting write-access to my mounted volumes (read access works fine). I will give it a closer look tomorrow.

About 2.: We can discuss those points, but in fact everything you do in that container is kept (I do not know if that was changed recently as well). When you kill the container or it exits normally its data is kept in the virtual file system of that container (not the image), and you can restart this container by finding its name or id via `docker ps -a`, and all created data will still be there. In fact this means if you often start new containers via `docker run image` (without the -rm option), and create data within the container you need to clean up every now and then (`docker container prune` removes all finished containers). From the docker documentation: `By default a container’s file system persists even after the container exits.` I agree that it is bad practice for us to create data outside of the mounted volume, because it might invisibly fill up user's hard drives, but it is not lost, and one might for example also create a workflow that only mounts a volume read-only, and then manually copy out the data. This was in fact my first workflow description, which is why the `docker container prune` command further down was left over (it needs to be removed now).

spikebike replied:
Looks good to me, the ideal container depends on the use case.  So far I've come up with:
1) runtime container - small as possible, good for tutorials, filing bug reports, ruling out that a problem isn't specific to a custom build, and potentially even production runs.
2) full source code/build from scratch container - good for CI testing, helping developers get started, allowing users to tune parameters, etc.  Sneaky sysadmins might even peak at the build scripts to clarify exactly how a release is built.

In either case it's pretty straight forward to just write what a user expects to their home directory.  That way input and output files could be kept, even if the user ditches the containers and upgrades to newer versions or builds from scratch without the containers.

How to handle dependencies is an interesting question.  Sure in an ideal world Deal.II would have a source code container that built a release into a runtime container.  Then the Aspect-devel container would use the deal.II runtime to build into a aspect-runtime container. 

For the file/permission issues I've use tinkered with entrypoint.sh and passing in the UID, that way the container can write to my homedir with my user ID.  My testing without it results in mounting my directory inside a docker container and then having new files owned by root when created.

tjhei commented:
> Do we want to offer this option? This should be answered after discussing the remaining questions.

I would say yes. It is good to provide options. I am somewhat scared (maybe unfounded) that doing "docker run" is easyt like you say, but that it is easy to mess things up. Second, I would prefer a solution that contains an easy way forward if you want to do more things (compile plugins, contribute, ...). With the current solution somebody might start editing code and lose the source, because it is not in the mounted directory for example.

> Is the current setup a reasonable one? Should we rely on the deal.II docker files, or offer our own solution? Building on deal.II requires less work on our side, but as far as I understood the dealii files are not officially supported either, so they might change, or be removed at some point.

They won't go away, but they are work in progress. I think it make sense to use them.

> In the current form I only build ASPECT in release mode.

I think debug mode (or providing debug and release executables) is preferable. You also delete the build directory, which I wouldn't do. At least do a ``make clean`` and keep the makefiles around.

gassmoeller commented:
I agree that we should provide debug and release. I removed the build folder to keep the image small, but I can take a look how much `make clean` reduces the size of the build folder. And I mostly moved the binary, because I wanted to remove the build folder, but if we keep it, I can as well create a copy or a symlink.
About the way forward: I also thought about writing a section in the manual about using the deal.II container to set up your own aspect source folder (e.g. mounting your host's aspect source folder into the container, and compile and run your models inside the container). This could be a good option for many advanced users, who would like to develop on their host system, and run inside the container.

tjhei commented:
> (or is this by accident because dealii and my host user happen to have the same uid?).

Yes, if dealii inside and your user outside have the same numerical uid, it works. You get lucky because your local user was the first one created on your machine and you have a similar OS so both are id 1000. There is some uid mapping stuff planned in docker, but that needs additional kernel support and I don't think that works yet.

tjhei commented:
> When you kill the container or it exits normally its data is kept in the virtual file system

Yes, of course. But a user has to know this and figure out how to find/restart a container.

spikebike commented:
+tjhei With a source build it is possible that changes could be lost.   Maybe we should save the state (diffs, git commits, or similar) for any changes on each build.  Of course those would go in the user's homedir, not inside the container.  Encouraging users to fork it and making it easier for users to submit pull requests seems ideal.  Maybe even the preferred method for pull requests.

tjhei commented:
> For the file/permission issues I've use tinkered with entrypoint.sh and passing in the UID, that way the container can write to my homedir with my user ID. My testing without it results in mounting my directory inside a docker container and then having new files owned by root when created.

I have seen this work-around but it is also not great from a usability standpoint (edit/visualize files owned by root from outside). Our current approach (hoping that the two uids match) is nicer when it works.

tjhei commented:
> Maybe we should save the state (diffs, git commits, or similar) for any changes on each build.

Well, you can't do that easily because you can't be sure that any folder is mounted from outside.

As I said: I am somewhat worried that the whole permission/mounting mess when using docker containers is difficult to teach and understand for inexperienced users.

gassmoeller commented:
I updated the dockerfile and added a debug executable. Still, I would prefer to remove the build directories after compiling, since even with `make clean` each contain at least 100MB of additional data that will likely never be necessary. In the current configuration I see the docker container more as a packaged *executable* than something that is supposed to act as a development environment. I feel that for development it would be more useful to mount an external source directory into the deal.II container, and I added a description of how to do this to the manual. Would you agree with this approach @tjhei?

bangerth commented:
Removing the object files seems reasonable as long as the rest of the cmake setup is preserved. If someone wants to make changes, they can then just type `make` to re-create the object files.

If you want to avoid people making changes to the sources (because you're afraid that they get lost upon shut-down of the image), why not just make everything in the source directory read-only?

tjhei commented:
>> each contain at least 100MB of additional data that will likely never be necessary

I don't think 100mb make a big difference and, like Wolfgang said, having the build setup is useful.

gassmoeller commented:
Well then lets not argue about 100~MB more or less. I modified the Dockerfile as requested. I also wrote the documentation of the Virtual machine option.

One open question would be under which name we distribute the Docker container. Will it be `geodynamics/aspect`? I suppose that would be most consistent. 
@tjesser-ucdavis-edu I see that there is a geodynamics/pylith container online, would you be the person to ask to upload an ASPECT image build from the Dockerfile included in this PR?

tjhei commented on docker/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+          -DDEAL_II_DIR=$HOME/deal.II-install \
+          .. && \
+    make -j4 >log 2>&1 && \
+    mv aspect $HOME/aspect/aspect-debug && \
```
rename to ``aspect`` not ``aspect-debug``??

tjhei commented on docker/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+    make -j4 >log 2>&1 && \
+    mv aspect $HOME/aspect/aspect-debug && \
+    make clean
+
```
add ``ENV ASPECT_DIR /home/bob/aspect/build-debug``

tjhei commented on docker/build.sh:
```diff
@@ -0,0 +1,10 @@
+# Note: This container is build from the developer version on Github, it does not use
+# the local ASPECT folder.
+
+docker build -t geodynamics/aspect .
```
just use ``gassmoeller/aspect`` for now?

gassmoeller commented:
I have addressed the comments and pushed the new image to docker hub. Is this ready for a merge?

tjhei commented:
I am happy with this. Any final comments, @bangerth ?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+
+\begin{enumerate}
+\item Create your \aspect{} input file in a folder of your choice (possibly also containing any input data that is required by your model) and navigate in a terminal into that directory.
+\item Run the docker image and mount the current directory as a read-only volume into the docker container\footnote{Note that it is possible to mount a directory as writeable into the container. However it is often associated with file permission conflicts between the host system and the container. Therefore, we recommend this slightly more cumbersome, but also more reliable workflow.}. 
```
However it is... -> However, this is often...

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+\begin{enumerate}
+\item Create your \aspect{} input file in a folder of your choice (possibly also containing any input data that is required by your model) and navigate in a terminal into that directory.
+\item Run the docker image and mount the current directory as a read-only volume into the docker container\footnote{Note that it is possible to mount a directory as writeable into the container. However it is often associated with file permission conflicts between the host system and the container. Therefore, we recommend this slightly more cumbersome, but also more reliable workflow.}. 
+Make sure your parameter file specifies a model output directory \textbf{other} than the input directory, e.g. \texttt{/home/dealii/aspect/model\_output}. When you have started the container run the aspect model inside the container. Note that there are two \aspect{} executables in the work directory of the container: texttt{aspect} and \texttt{aspect-release}. For a discussion of the different v ...
```
\textbf{other} -> \textit{other}

bangerth replied:
texttt{aspect}: missing backslash

bangerth replied:
of the different version -> of the different versions

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+Make sure your parameter file specifies a model output directory \textbf{other} than the input directory, e.g. \texttt{/home/dealii/aspect/model\_output}. When you have started the container run the aspect model inside the container. Note that there are two \aspect{} executables in the work directory of the container: texttt{aspect} and \texttt{aspect-release}. For a discussion of the different v ...
+
+\begin{verbatim}
+docker run -it -v "$(pwd):/home/dealii/aspect/model_input:ro" \
```
This starts a bit unmotivated. Maybe prefix it with "To sum up, the steps you will want to execute are:..."

Also, this is in essence a shell script. Instead of `verbatim`, use the `lstslisting` environment and mark it as a `ksh` script -- other examples of this abound in the manual.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+\item After the model has finished (or during the model run if you want to check intermediate results) copy the model output out of the container into your current directory. For this you need to find the name or ID of the docker container by running \texttt{docker ps -a} in a separate terminal first. Look for the most recently started container to identify your current \aspect{} container.
+
+\begin{verbatim}
+docker ps -a # Find the name of the running / recently closed container in the output
```
same here -- introduce the command block with a sentence

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+docker cp CONTAINER_NAME:/home/dealii/aspect/model_output .
+\end{verbatim}
+
+\item The output data is saved inside your container even after the computation finishes and even when you stop the container. After you have copied out the data of the container you should therefore delete the container to avoid duplication of output data. Even after deleting you will always be able to start a new container from the downloaded image following step 2. Deleting the finished contai ...
```
-> After you have copied the data out of the container

bangerth replied:
Instead of hiding the warning in a parenthetical remark, don't we have a `\warning` or `\note` environment?

I don't think I know what you mean by "only" in "that you would like to keep only use"

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+\item The output data is saved inside your container even after the computation finishes and even when you stop the container. After you have copied out the data of the container you should therefore delete the container to avoid duplication of output data. Even after deleting you will always be able to start a new container from the downloaded image following step 2. Deleting the finished contai ...
+
+\begin{verbatim}
+docker container prune
```
intro sentence

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+\begin{verbatim}
+docker pull dealii/dealii:v8.5.pre.4-gcc-mpi-fulldepsmanual-debugrelease
+docker run -it -v "$(pwd):/home/dealii/aspect:ro" \
+  dealii/dealii:v8.5.pre.4-gcc-mpi-fulldepsmanual-debugrelease bash
```
we'll have to update this with every release. If we can't use an unversioned deal.II container, we ought to put updating this line into the release task list

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+
+Inside of the container you now find a read-only \aspect{} directory that contains your modified source code. You can compile and run a model inside the container, e.g. in the following way:
+
+\begin{verbatim}
```
like above, use the lstlisting environment. same for all other places here where you show command line commands.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+
+\subsubsection{Installing VM software and setting up the virtual machine}
+
+We are providing an experimental virtual image for a fully configured version of \aspect{}. For this, you will need to install VirtualBox(\url{http://www.virtualbox.org/}) on your machine, and then import a virtual machine image that can be downloaded from \url{http://www.math.clemson.edu/~heister/dealvm/}. Note, however, that the machine image is several gigabytes in size and downloading will ta ...
```
I'm not fond of the "We are" way of writing things (who is "we" anyway). I would say "The \aspect{} project is providing..."

bangerth replied:
space after "VirtualBox"

bangerth commented:
Nice! Just a few small nitpicks. OK to merge after fixing.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -2268,10 +2268,152 @@ \subsection{Particles}
+\begin{enumerate}
+\item Virtual Machine (\aspect{} beginner and tutorial participant): The virtual machine image provides a fully prepared user environment that contains installations of \aspect{}, all required libraries, and visualization software on top of a full Linux environment. This way beginning users and tutorial participants can work in a unified  environment, thus minimizing installation time and technic ...
+
+\item Docker Container (advanced user with no need to configure / change the underlying libraries, possibly changing parts of \aspect): Docker containers are lightweight packages that only encapsulate the minimal dependencies to run an application like \aspect{} on top of the host operating system. They allow easy installation and usage of \aspect{} in a unified environment, while relying on the  ...
```
I removed the spaces. 
But you do not need to run a linux box. Docker is available for Mac and Windows as well. The minor difference is that on Windows Docker runs a minimal linux virtual machine that executes the containers. But the integration of this machine has become so effective that there is nearly no difference to running the containers natively. This is why I do not mention it, it would only confuse users.

bangerth replied:
Oh, I see. But then maybe it's useful to say that running it on Windows the result looks like a virtual machine in that it provides a unix-like environment. Or is that not true either?

gassmoeller replied:
But it does not even look like a virtual machine. You use the Windows shell with commands identical to the commands in a unix shell. That docker internally redirects these commands to a unix virtual machine is not visible at all. The only break is that once the container is started with `docker run -it -v "$(pwd):/home/dealii/aspect/model_input:ro" \
gassmoeller/aspect:latest bash` one switches to a new linux bash shell inside the container, but that is the same as in the linux environment.

bangerth replied:
Ah, nice. That's actually a far nicer approach than I thought it would be on Windows!
